utils/git.go

```diff
@@ -19,20 +22,26 @@ func GitClone(remoteURL string) (string, error) {
 		return "", err
 	}
 
-	clone := cloneArgs(remoteURL, root)
+	u, err := url.Parse(remoteURL)
+	if err != nil {
+		return "", err
+	}
 
-	if output, err := exec.Command("git", clone...).CombinedOutput(); err != nil {
+	fragment := u.Fragment
+	clone := cloneArgs(u, root)
+
+	if output, err := git(clone...); err != nil {
 		return "", fmt.Errorf("Error trying to use git: %s (%s)", err, output)
 	}
 
-	return root, nil
+	return checkoutGit(fragment, root)
 }
 
-func cloneArgs(remoteURL, root string) []string {
+func cloneArgs(remoteURL *url.URL, root string) []string {
 	args := []string{"clone", "--recursive"}
-	shallow := true
+	shallow := len(remoteURL.Fragment) == 0
```

Comment:
yes. A shallow clone only gives you the last commit in the history, you cannot checkout a branch because your local repo doesn't know what that is. You need to run a `git fetch` if you want to do that, which is exactly the same as running a full clone.

So the logic is as follow:

1. Shallow clone by default because you don't want to do anything with the git history yet.
2. Full clone because we need to manipulate the repo to checkout a reference.